plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/AggregationQueryPageSource.java

```diff
@@ -0,0 +1,259 @@
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregation;
+import org.elasticsearch.search.aggregations.metrics.stats.Stats;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.trino.plugin.elasticsearch.ElasticsearchQueryBuilder.buildAggregationQuery;
+import static io.trino.plugin.elasticsearch.ElasticsearchQueryBuilder.buildSearchQuery;
+import static java.util.Objects.requireNonNull;
+
+public class AggregationQueryPageSource
+        implements ConnectorPageSource
+{
+    private long readTimeNanos;
+    private final List<Decoder> decoders;
+    private final BlockBuilder[] columnBuilders;
+    private final List<ElasticsearchColumnHandle> columns;
+    private final ElasticsearchClient client;
+    private final ElasticsearchTableHandle table;
+    private final ElasticsearchSplit split;
+    private final int pageSize;
+    // the after parameter that composite aggregation used to fetch by page
+    private Optional<Map<String, Object>> after;
+    private boolean fetched;
+    private static final SearchHit noUsed = new SearchHit(0);
+
+    public AggregationQueryPageSource(
+            ElasticsearchClient client,
+            ElasticsearchTableHandle table,
+            ElasticsearchSplit split,
+            List<ElasticsearchColumnHandle> columns,
+            int pageSize)
+    {
+        requireNonNull(client, "client is null");
+        requireNonNull(table, "table is null");
+        requireNonNull(split, "split is null");
+        requireNonNull(columns, "columns is null");
+        this.client = client;
+        this.table = table;
+        this.split = split;
+        this.pageSize = (int) table.getLimit().stream().filter(limit -> limit < pageSize).findFirst().orElse(pageSize);
+        this.columns = columns;
+        this.after = Optional.empty();
+        columnBuilders = columns.stream()
+                .map(ElasticsearchColumnHandle::getType)
+                .map(type -> type.createBlockBuilder(null, 1))
+                .toArray(BlockBuilder[]::new);
+        decoders = createDecoders(columns);
+    }
+
+    @Override
+    public long getCompletedBytes()
+    {
+        return 0;
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        return readTimeNanos;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return fetched && after.isEmpty();
+    }
+
+    @Override
+    public Page getNextPage()
+    {
+        long start = System.nanoTime();
+        SearchResponse searchResponse = client.beginSearch(
+                split.getIndex(),
+                split.getShard(),
+                buildSearchQuery(table.getConstraint().transform(ElasticsearchColumnHandle.class::cast), table.getQuery()),
```

Comment:
This should be done once in the constructor instead of in every call of `getNextPage()`.

Also, `TupleDomain.transform()` is deprecated, so change it to `transformKeys`.